osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/factories/InvocationCounterManagerTest.java

```diff
@@ -93,4 +101,41 @@ void initializesInvocationCounterForDevice() throws Exception {
     assertThat(this.device.getVersion()).isGreaterThan(this.initialDeviceVersion);
     assertThat(this.device.getInvocationCounter()).isEqualTo(invocationCounterValueOnDevice);
   }
+
+  @Test
+  void attemptToLowerInvocationCounterForDeviceLogsErrorAndThrowsException() throws Exception {
+    final Logger invocationManagerLogger =
+        (Logger) LoggerFactory.getLogger(InvocationCounterManager.class);
+    final ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+    final List<ILoggingEvent> logsList = listAppender.list;
+    final long invocationCounterValueOnDevice = 1;
+    final DlmsConnectionManager connectionManager = mock(DlmsConnectionManager.class);
+    final DataObject dataObject = DataObject.newUInteger32Data(invocationCounterValueOnDevice);
+
+    when(this.dlmsHelper.getAttributeValue(
+            eq(connectionManager), refEq(ATTRIBUTE_ADDRESS_INVOCATION_COUNTER_VALUE)))
+        .thenReturn(dataObject);
+
+    listAppender.start();
+    invocationManagerLogger.addAppender(listAppender);
+
+    final FunctionalException thrownException =
+        catchThrowableOfType(
+            () ->
+                this.manager.initializeWithInvocationCounterStoredOnDeviceTask(
+                    this.device, connectionManager),
+            FunctionalException.class);
+
+    assertThat(logsList.size() > 0).isTrue();
+    assertThat(
+            logsList
+                .get(0)
+                .toString()
+                .contains("[ERROR] Attempt to lower invocationCounter of device"))
+        .isTrue();
```

Comment:
As stated, I am not a fan of assertions on logging, but if you do so, I think the following would improve things a bit:
```java
    final ILoggingEvent loggingEvent = logsList.get(0);
    assertThat(loggingEvent).asString().contains("Attempt to lower invocationCounter of device");
    assertThat(loggingEvent.getLevel()).isEqualTo(Level.ERROR);
```
What I believe to be nicer in this alternative compared to the current code is that it takes out the "[ERROR]" part, and replaces it by an explicit assert on the level. The formatting of the word "ERROR" with the square brackets could change when using different logging configuration, and you would probably not want that to break the test. By using `contains` on the String assertion, you would get a nicer error message than with `isTrue` (boolean assertion) on the result of `String.contains`.

Reply:
yes, should have done this in the first place. It is updated now.